azure/services/privatedns/privatedns.go

```diff
@@ -57,178 +71,103 @@ func (s *Service) Name() string {
-				continue
-			}
-			// Link each virtual network.
-			log.V(2).Info("creating a virtual network link", "virtual network", linkSpec.VNetName, "private dns zone", zoneSpec.ZoneName)
-			link := privatedns.VirtualNetworkLink{
-				VirtualNetworkLinkProperties: &privatedns.VirtualNetworkLinkProperties{
-					VirtualNetwork: &privatedns.SubResource{
-						ID: to.StringPtr(azure.VNetID(s.Scope.SubscriptionID(), linkSpec.VNetResourceGroup, linkSpec.VNetName)),
-					},
-					RegistrationEnabled: to.BoolPtr(false),
-				},
-				Location: to.StringPtr(azure.Global),
-				Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-					ClusterName: s.Scope.ClusterName(),
-					Lifecycle:   infrav1.ResourceLifecycleOwned,
-					Additional:  s.Scope.AdditionalTags(),
-				})),
-			}
-			err = s.client.CreateOrUpdateLink(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, linkSpec.LinkName, link)
-			if err != nil {
-				return errors.Wrapf(err, "failed to create virtual network link %s", linkSpec.LinkName)
-			}
-			log.V(2).Info("successfully created virtual network link", "virtual network", linkSpec.VNetName, "private dns zone", zoneSpec.ZoneName)
-		}
-		// Create the record(s).
-		for _, record := range zoneSpec.Records {
-			log.V(2).Info("creating record set", "private dns zone", zoneSpec.ZoneName, "record", record.Hostname)
-			set := privatedns.RecordSet{
-				RecordSetProperties: &privatedns.RecordSetProperties{
-					TTL: to.Int64Ptr(300),
-				},
-			}
-			recordType := converters.GetRecordType(record.IP)
-			if recordType == privatedns.A {
-				set.RecordSetProperties.ARecords = &[]privatedns.ARecord{{
-					Ipv4Address: &record.IP,
-				}}
-			} else if recordType == privatedns.AAAA {
-				set.RecordSetProperties.AaaaRecords = &[]privatedns.AaaaRecord{{
-					Ipv6Address: &record.IP,
-				}}
-			}
-			err := s.client.CreateOrUpdateRecordSet(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, recordType, record.Hostname, set)
-			if err != nil {
-				return errors.Wrapf(err, "failed to create record %s in private DNS zone %s", record.Hostname, zoneSpec.ZoneName)
-			}
-			log.V(2).Info("successfully created record set", "private dns zone", zoneSpec.ZoneName, "record", record.Hostname)
-		}
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	zoneSpec, links, records := s.Scope.PrivateDNSSpec()
+	if zoneSpec == nil {
+		return nil
+	}
+
+	managed, err := s.reconcileZone(ctx, zoneSpec)
+	if managed {
+		s.Scope.UpdatePutStatus(infrav1.PrivateDNSZoneReadyCondition, serviceName, err)
+	}
+	if err != nil {
+		return err
+	}
+
+	managed, err = s.reconcileLinks(ctx, links)
+	if managed {
+		s.Scope.UpdatePutStatus(infrav1.PrivateDNSLinkReadyCondition, serviceName, err)
+	}
+	if err != nil {
+		return err
 	}
-	return nil
+
+	err = s.reconcileRecords(ctx, records)
+	s.Scope.UpdatePutStatus(infrav1.PrivateDNSRecordReadyCondition, serviceName, err)
```

Comment:
I think we only want to update the put status if we actually have specs. In this case, we could only call `UpdatePutStatus` if the number of specs > 0. The same goes for the delete status too.

Reply:
Okay sounds good. In that case I think it LGTM.